File: backtracking/permutations.js
var permute = function(nums) {
    const backtrack = curr => {
        if (curr.length == nums.length)
        {
            ans.push([...curr]);
            return;
        }

        for (const num of nums)
        {
            if (!curr.includes(num))
            {
                curr.push(num);
                backtrack(curr);
                curr.pop();
            }
        }
    }

    let ans = [];
    backtrack([]);
    return ans;
};

File: backtracking/permutations.py
def permute(nums: list[int]) -> list[list[int]]:
    def backtrack(curr):
        if len(curr) == len(nums):
            ans.append(curr[:])
            return

        for num in nums:
            if num not in curr:
                curr.append(num)
                backtrack(curr)
                curr.pop()

    ans = []
    backtrack([])
    return ans
